File: node/engines/ddg.js
const { getBrowser } = require("../_browser");

async function fetchDDG(query, n = 10) {
  const browser = await getBrowser();
  const page = await browser.newPage();
  try {
    await page.goto(
      `https://duckduckgo.com/?q=${encodeURIComponent(query)}&iax=images&ia=images`,
      { waitUntil: "networkidle" }
    );
    return await page.$$eval(
      'img[src*="external-content.duckduckgo.com"]',
      (imgs, maxN) => {
        const urls = [];
        const seen = new Set();
        for (const img of imgs) {
          try {
            const src = img.getAttribute("src") || "";
            const fullSrc = src.startsWith("//") ? "https:" + src : src;
            const u = new URL(fullSrc).searchParams.get("u");
            if (u && u.startsWith("http") && !seen.has(u)) {
              seen.add(u);
              urls.push(u);
              if (urls.length >= maxN) break;
            }
          } catch {}
        }
        return urls;
      },
      n
    );
  } finally {
    await page.close();
  }
}

module.exports = { fetchDDG };

File: imgsearch_api/engines/ddg.py
from urllib.parse import quote_plus, urlparse, parse_qs, unquote
from .._fetcher import get_fetcher


def fetch_ddg(query: str, n: int = 10) -> list[str]:
    # Load the image search page with browser (JS required for results)
    page = get_fetcher().fetch(
        f"https://duckduckgo.com/?q={quote_plus(query)}&iax=images&ia=images",
        headless=True,
        network_idle=True,
    )

    urls: list[str] = []
    seen: set[str] = set()

    # DDG thumbnails are served via their CDN proxy:
    # https://external-content.duckduckgo.com/iu/?u=REAL_ENCODED_URL&...
    # Extract the real URL from the `u` query param.
    for img in page.css("img[src*='external-content.duckduckgo.com']"):
        src = img.attrib.get("src", "")
        if not src:
            continue
        params = parse_qs(urlparse(src).query)
        real_url = unquote(params.get("u", [""])[0])
        if real_url and real_url.startswith("http") and real_url not in seen:
            seen.add(real_url)
            urls.append(real_url)
            if len(urls) >= n:
                break

    return urls
